### PR TITLE
Kiali 1205 [1.0] graph, update response_time appender

### PR DESCRIPTION
**Pull request by jshaughn**

### CODE REVIEW

**Comment by jshaughn** [comment]
This PR is blocked by https://github.com/kiali/kiali/pull/361

**Comment by jshaughn** [comment]
@jmazzitelli I think this appender is working, give it a spin if you get the chance.

File: graph/appender/response_time.go [hunk 1]
```diff
@@ -48,60 +50,98 @@ func (a ResponseTimeAppender) appendGraph(trafficMap graph.TrafficMap, namespace
 	}
 	log.Warningf("Generating responseTime using quantile [%.2f]", quantile)
 
-	// query prometheus for the responseTime info in two queries. The first query gathers responseTime for
-	// requests originating outside of the namespace...
-	namespacePattern := fmt.Sprintf(".*\\\\.%v\\\\..*", namespace)
-	query := fmt.Sprintf("histogram_quantile(%.2f, sum(rate(%s{source_service!~\"%v\",destination_service=~\"%v\",response_code=\"200\"}[%vs])) by (%s))",
+	// query prometheus for the responseTime info in three queries:
+	// 1) query for responseTime originating from "unknown" (i.e. the internet)
+	groupBy := "le,source_workload_namespace,source_workload,source_app,source_version,destination_service_namespace,destination_service_name,destination_workload,destination_app,destination_version"
```
Comment (jmazzitelli) [inline graph/appender/response_time.go]: what is "le" ?

File: graph/appender/response_time.go [hunk 2]
```diff
@@ -48,60 +50,98 @@ func (a ResponseTimeAppender) appendGraph(trafficMap graph.TrafficMap, namespace
...
 		}
 
-		key := fmt.Sprintf("%s %s %s %s", sourceSvc, sourceVer, destSvc, destVer)
+		// For source-side failures (Fault injection, open circuit breakers, etc) the
+		// destination_workload will be unknown (because the request never made it to the
+		// destination proxy.  But we still want to record the request in some
+		// way. So, depending on the graph type and labeling, assign the destination
+		// service name (i.e. the requested service) to the most applicable destination field.
+		// note: this code is duplicated in graph.go for any changes/fixes
+		if string(destApp) == graph.UnknownApp {
```
Comment (jmazzitelli) [inline graph/appender/response_time.go]: the comment says "...the destination_workload will be unknown..." but yet the if-stmt checks if destApp is unknown. Shouldn't this be checking if "destWl" is unknown?

File: handlers/graph.go [hunk 3]
```diff
@@ -212,13 +212,14 @@ func populateTrafficMap(trafficMap graph.TrafficMap, vector *model.Vector, o opt
 		// destination proxy.  But we still want to record the request in some
 		// way. So, depending on the graph type and labeling, assign the destination
 		// service name (i.e. the requested service) to the most applicable destination field.
+		// note: this code is duplicated in response_time.go for any changes/fixes
 		if destApp == graph.UnknownApp {
```
Comment (jmazzitelli) [inline handlers/graph.go]: shouldn't this check if destWl is unknown? see earlier comment

File: graph/appender/response_time.go (hunk 1 above)
Comment (jshaughn) [reply graph/appender/response_time.go]: "le" is a prometheus thing, it's a "bucket label" when working with histogram quantiles. See https://prometheus.io/docs/practices/histograms/#quantiles.

File: graph/appender/response_time.go (hunk 2 above)
Comment (jshaughn) [reply graph/appender/response_time.go]: The power of comments.  Good catch.  DestApp will also be unknown, but that could also happen in app labels were not being used.  So for our current use cases this works as is, but it does need to change to cover all cases. I'll make the update.

File: handlers/graph.go (hunk 3 above)
Comment (jshaughn) [reply handlers/graph.go]: Right, same problem, see comment about duplicated code :)

Comment (jmazzitelli) [reply handlers/graph.go]: can refactor this into a utility function somewhere and reuse so we don't duplicate bugs... er duplicate code? :)

Comment (jshaughn) [reply handlers/graph.go]: Probably, I actually stared at this for a while last night but couldn't decide on a good place to put the shared code. So you get this instead.  Maybe I'll move it later but not now.

**Pull request merged**